airflow/models/taskinstance.py

```diff
@@ -1204,10 +1204,12 @@ def check_and_change_state_before_execution(
             # For reporting purposes, we report based on 1-indexed,
             # not 0-indexed lists (i.e. Attempt 1 instead of
             # Attempt 0 for the first attempt).
-            # Set the task start date. In case it was re-scheduled use the initial
-            # start date that is recorded in task_reschedule table
+            # Set the task start date.
             self.start_date = timezone.utcnow()
-            if self.state == State.UP_FOR_RESCHEDULE:
+
+            # In case it was re-scheduled use the initial
+            # start date that is recorded in task_reschedule table
+            if hasattr(self.task, 'reschedule') and self.task.reschedule:
```

Comment:
The  `reschedule` property is part of `BaseSensorOperator` , but this is not the only kind of reschedulable operator.   Any operator can raise `AirflowRescheduleException` to  reschedule and may not have the `reschedule` property.  So I don't think we use this property  to infer what to do here.

Reply:
Ahh, good point!

Since any operator can raise `AirflowRescheduleException`, how would I distinguish between tasks that are being rescheduled? Would I need to query the `task_reschedule` table for all tasks? This would remove the original optimization though maybe it doesn't make sense to have it in the first place.

Reply:
> Since any operator can raise AirflowRescheduleException, how would I distinguish between tasks that are being rescheduled? Would I need to query the task_reschedule table for all tasks? This would remove the original optimization though maybe it doesn't make sense to have it in the first place.


with the caveat that i don't know this code all that well...

so, what you have pointed out is that `check_and_change_state_before_execution` never runs until after state is already flipped to `scheduled` by `schedule_tis`, right?  that makes sense.

i'm not seeing anything that's already there that would allow us to determine whether the TI is resuming from a reschedule or some other case.  i think any solution is likely to be rather involved.

e.g. one option would be to split up the task states.  we could think of some of the states as being different "kinds" of state. some overlap.  e.g. you can be "up for retry" from the moment a task fails until the moment it is actually running again.  meanwhile the task can cycle through many other "states" in this period of time.

Along these lines, we could think of these as one kind of state (call it "execution state"):

```
none
scheduled
queued
running
shutdown
restarting
paused  <-- just made this up
done <-- made this up too
```

To this we might add a category called "resolutions"

```
success
failed
skipped
upstream_failed
removed
```

Tasks in a `done` state would have _one_ of the above resolutions.

Then finally we could have one more attribute, named _what_ i'm not sure -- but it could have these values:
```
up_for_retry
up_for_reschedule
deferred
```

Essentially these are the _reasons_ why the task would be in a "paused" state.

By splitting things up in this way you would know, when you are flipping a task from queued to running for example, whether it's a resuming after a retry, or a reschedule, or a deferral.

But, I doubt there would be appetite to do something like this.

Maybe @ashb or @kaxil would have other ideas?